test/test_nestedtensor.py

```diff
@@ -3951,6 +3954,142 @@ def in_proj(input_packed, qkv_linear=qkv_linear):
+
+        with self.assertRaisesRegex(AssertionError, "has been mutated"):
+            registry.maybe_set_metadata(vec, "foo", "bar")
+
+        with self.assertRaisesRegex(AssertionError, "has been mutated"):
+            registry.get_metadata(vec, "foo")
+
+    def test_return_canonical_nested_int_and_vec(self):
+        # Add another vec to the same equiv set returns the canonical nested int
+        registry = torch.nested._internal.nested_tensor.NestedIntRegistry()
+        vec = torch.tensor([1, 2, 3])
+        vec_clone = vec.clone()
+        nested_int = registry.maybe_create(vec)
+        nested_int2 = registry.maybe_create(vec_clone, equiv_set_from=nested_int.node.nested_int_vec())
+        self.assertIs(nested_int2, nested_int)
+
+    def test_ownership(self):
+        registry = torch.nested._internal.nested_tensor.NestedIntRegistry()
+        vec = torch.tensor([1, 2, 3])
+        nested_int = registry.maybe_create(vec)
+
+        # Vec does NOT keep the nested int alive
+        ref = weakref.ref(nested_int)
+        with disable_gc():
+            # There are no reference cycles
+            del nested_int
+            self.assertIsNone(ref())
+
+        # Nested int DOES keep the vec alive
+        nested_int = registry.maybe_create(vec)
+        ref = weakref.ref(vec)
+        del vec
+        self.assertIs(ref(), nested_int.node.nested_int_vec())
+
+    def test_changing_canonical_nested_int_and_vec(self):
+        registry = torch.nested._internal.nested_tensor.NestedIntRegistry()
+        vec = torch.tensor([1, 2, 3])
+        nested_int = registry.maybe_create(vec)
+        equiv_set_id = nested_int.node.nested_int()
+        del nested_int
+        nested_int2 = registry.maybe_create(vec)
+        # As long as the vec stays alive, the equiv_set_id should stay the same
+        self.assertEqual(nested_int2.node.nested_int(), equiv_set_id)
+
+        vec_clone = vec.clone()
+        nested_int3 = registry.maybe_create(vec_clone, equiv_set_from=nested_int2.node.nested_int_vec())
+        self.assertIs(nested_int3, nested_int2)
+        ref = weakref.ref(vec)
+        del vec, nested_int2, nested_int3
+        # vec is dead because there are no more nested ints keeping it alive
+        self.assertIsNone(ref())
+
+        nested_int4 = registry.maybe_create(vec_clone)
+        # The equiv_set_id should not have changed because at least one vec
+        # in the equiv set is still alive: vec_clone
+        self.assertEqual(nested_int4.node.nested_int(), equiv_set_id)
+        # The new canonical vec is vec_clone
+        self.assertIs(nested_int4.node.nested_int_vec(), vec_clone)
+
+        self.assertTrue(len(registry._equiv_sets) == 1)
+        with disable_gc():
+            del vec_clone, nested_int4
+            self.assertTrue(len(registry._equiv_sets) == 0)
+
+    def test_custom_equiv_set(self):
+        # A equiv set is "custom" if its canonical int was created using a
+        # custom ctor_fn.
```

Comment:
what's the significance and purpose of a "custom" equiv set in practice?

Reply:
The main use is so that during compile, if I need a symbolic nested int, the creation of that can reuse the same logic as eager, for example enforcing that there is a single canonical vec/nested int